File: 19-Binary-Tree-Part-III/python/07_children_sum_property_in_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def checkChildrenSum(self, root: Optional[TreeNode]) -> bool:
        if not root or (not root.left and not root.right):
            return True
        left_val = root.left.val if root.left else 0
        right_val = root.right.val if root.right else 0
        if root.val != left_val + right_val:
            return False
        return self.checkChildrenSum(root.left) and self.checkChildrenSum(root.right)

File: 19-Binary-Tree-Part-III/java/07-ChildrenSumPropertyInBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean checkChildrenSum(TreeNode root) {
        if (root == null || (root.left == null && root.right == null)) return true;
        int leftVal = root.left != null ? root.left.val : 0;
        int rightVal = root.right != null ? root.right.val : 0;
        if (root.val != leftVal + rightVal) return false;
        return checkChildrenSum(root.left) && checkChildrenSum(root.right);
    }
}
